packages/block-library/src/link-preview/edit.js

```diff
@@ -0,0 +1,136 @@
+/**
+ * WordPress dependencies
+ */
+import { __, _x } from '@wordpress/i18n';
+import { useState, useEffect } from '@wordpress/element';
+import { useBlockProps, BlockControls } from '@wordpress/block-editor';
+import { __experimentalFetchUrlData } from '@wordpress/core-data';
+import {
+	Placeholder,
+	Spinner,
+	Button,
+	ToolbarGroup,
+	ToolbarButton,
+} from '@wordpress/components';
+import { link, edit } from '@wordpress/icons';
+
+export default function LinkPreviewEdit( props ) {
+	const {
+		attributes: { url, title, icon, image },
+		setAttributes,
+	} = props;
+	const [ isFetching, setIsFetching ] = useState( false );
+	const [ isEditingUrl, setIsEditingUrl ] = useState( ! url );
+	const [ urlValue, setURLValue ] = useState( '' );
+
+	const blockProps = useBlockProps( {
+		href: url,
+		className: image ? 'has-image' : undefined,
+		onClick: isEditingUrl
+			? undefined
+			: ( event ) => {
+					event.preventDefault();
+			  },
+	} );
+
+	useEffect( () => {
+		if ( url && ! title && ! icon && ! image ) {
+			setIsFetching( true );
+			__experimentalFetchUrlData( url )
+				.then( ( data ) => {
+					setAttributes( data );
+				} )
+				.finally( () => {
+					setIsFetching( false );
+				} );
+		}
+	}, [] );
+
+	if ( isEditingUrl ) {
+		return (
+			<div { ...blockProps }>
+				<Placeholder
+					icon={ link }
+					label={ __( 'URL' ) }
+					instructions={ __(
+						'Paste a link to the content you want to display on your site.'
+					) }
+				>
+					<form
+						onSubmit={ () => {
+							setAttributes( {
+								url: urlValue,
+								title: '',
+								icon: '',
+								image: '',
+								description: '',
+							} );
+							setIsFetching( true );
+							__experimentalFetchUrlData( urlValue )
+								.then( ( data ) => {
+									setAttributes( data );
+									setIsEditingUrl( false );
+								} )
+								.finally( () => {
+									setIsFetching( false );
+								} );
+						} }
+					>
+						<input
+							type="url"
+							value={ urlValue }
+							className="components-placeholder__input"
+							aria-label={ __( 'URL' ) }
+							placeholder={ __( 'Enter URL to embed here…' ) }
+							onChange={ ( event ) => {
+								setURLValue( event.target.value );
+							} }
+						/>
+						<Button variant="primary" type="submit">
+							{ _x( 'Embed', 'button label' ) }
+						</Button>
+					</form>
+				</Placeholder>
+			</div>
+		);
+	}
+
+	return (
+		<a { ...blockProps }>
+			<BlockControls>
+				<ToolbarGroup>
+					<ToolbarButton
+						className="components-toolbar__control"
+						label={ __( 'Edit URL' ) }
+						icon={ edit }
+						onClick={ () => {
+							setIsEditingUrl( true );
+							setURLValue( url );
+						} }
+					/>
+				</ToolbarGroup>
+			</BlockControls>
+
+			{ isFetching && <Spinner /> }
+
+			{ ! isFetching && (
```

Comment:
I pushed this change so we see a spinner whilst fetching rather than half of the preview.

We should also consider what to show if the rich url data is not available for the given link. Some kind of fallback for each attribute perhaps? Currently it could look like an error. 

Instead we should try to communicate to the user that we are missing some of the data.

Reply:
I think it's ok if either the icon or image is not available. Maybe we could make the title a requirement because every web page should have at least a title.

Reply:
...it _should_ but the endpoint is obviously struggling with certain sites. I guess all I'm suggesting is we provide a visual fallback for `icon` and `image` so the layout is retained. It's just a suggestion though. I should probably defer to @jasmussen on that one 🙇

Reply:
Made the title a requirement.